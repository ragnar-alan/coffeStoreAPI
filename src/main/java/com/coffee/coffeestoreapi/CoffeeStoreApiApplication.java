package com.coffee.coffeestoreapi;

import org.springframework.boot.SpringApplication;
import org.springframework.boot.autoconfigure.SpringBootApplication;

@SpringBootApplication
public class CoffeeStoreApiApplication {

    public static void main(String[] args) {
        SpringApplication.run(CoffeeStoreApiApplication.class, args);
    }

}
