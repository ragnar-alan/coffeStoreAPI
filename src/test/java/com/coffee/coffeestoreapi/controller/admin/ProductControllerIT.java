package com.coffee.coffeestoreapi.controller.admin;

import com.coffee.coffeestoreapi.BaseIT;
import io.restassured.RestAssured;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.springframework.boot.test.web.server.LocalServerPort;
import org.springframework.test.context.DynamicPropertyRegistry;
import org.springframework.test.context.DynamicPropertySource;
import org.testcontainers.containers.PostgreSQLContainer;
import org.testcontainers.junit.jupiter.Container;
import org.testcontainers.utility.DockerImageName;

import static org.approvaltests.JsonApprovals.verifyJson;

public class ProductControllerIT extends BaseIT {
    private static final DockerImageName IMAGE_NAME = DockerImageName
            .parse("postgres:16-alpine")
            .asCompatibleSubstituteFor("postgres");
    @Container
    static final PostgreSQLContainer<?> POSTGRES_CONTAINER = new PostgreSQLContainer<>(IMAGE_NAME)
            .withDatabaseName("testdb")
            .withUsername("testuser")
            .withPassword("testpass");

    @LocalServerPort
    private int port;

    static {
        POSTGRES_CONTAINER.start();
    }

    @DynamicPropertySource
    static void overrideProperties(DynamicPropertyRegistry registry) {
        registry.add("spring.datasource.url", POSTGRES_CONTAINER::getJdbcUrl);
        registry.add("spring.datasource.username", POSTGRES_CONTAINER::getUsername);
        registry.add("spring.datasource.password", POSTGRES_CONTAINER::getPassword);
    }

    @BeforeEach
    void setUp() {
        RestAssured.port = port;
    }

    @Test
    void testGetProductsShouldPass() {
        var result = RestAssured
                .given()
                .when()
                    .get("/api/v1/admin/products/list")
                .then()
                    .statusCode(200)
                    .extract()
                    .asString();
        verifyJson(result);
    }

    @Test
    void testGetProductShouldPass() {
        var result = RestAssured
                .given()
                .when()
                    .get("/api/v1/admin/products/{productId}", 1)
                .then()
                    .statusCode(200)
                    .extract()
                    .asString();
        verifyJson(result);
    }

    @Test
    void testGetProductShouldFail_dueToMissingResource() {
        RestAssured
                .given()
                .when()
                    .get("/api/v1/admin/products/{productId}", 534)
                .then()
                    .statusCode(404)
                    .extract()
                    .asString();
    }

// I wasn't able to get this test to pass. Somewhere I messed up the configuration maybe, because the other tests are fine.
// I will leave it commented out for now, but I will try to fix it later.
    /*@Test
    void testCreateProductShouldPass() throws IOException {
        var result = RestAssured
                .given()
                    .contentType(MediaType.APPLICATION_JSON_VALUE)
                    .body(getFileContents("mockRequests/mockCreateProductRequest.json"))
                .when()
                    .post("/api/v1/admin/products")
                .then()
                    .statusCode(201)
                    .extract()
                    .asString();
        verifyJson(result);
    }*/

    @Test
    void testGetMostPopularItemsShouldPass() {
        var result = RestAssured
                .given()
                .when()
                    .get("/api/v1/admin/products/most-popular")
                .then()
                    .statusCode(200)
                .extract()
                .asString();
        verifyJson(result);
    }
}
