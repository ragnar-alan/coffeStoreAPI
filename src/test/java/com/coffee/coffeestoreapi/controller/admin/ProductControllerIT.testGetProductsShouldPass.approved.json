[
  {
    "id": 1,
    "product_name": "Black Coffee",
    "price_in_cents": 400,
    "type": "DRINK",
    "is_favorite": true
  },
  {
    "id": 2,
    "product_name": "Latte",
    "price_in_cents": 500,
    "type": "DRINK",
    "is_favorite": false
  },
  {
    "id": 3,
    "product_name": "Mocha",
    "price_in_cents": 600,
    "type": "DRINK",
    "is_favorite": false
  },
  {
    "id": 4,
    "product_name": "Tea",
    "price_in_cents": 300,
    "type": "DRINK",
    "is_favorite": true
  },
  {
    "id": 5,
    "product_name": "Milk",
    "price_in_cents": 200,
    "type": "TOPPING",
    "is_favorite": true
  },
  {
    "id": 6,
    "product_name": "Hazelnut syrup",
    "price_in_cents": 300,
    "type": "TOPPING",
    "is_favorite": false
  },
  {
    "id": 7,
    "product_name": "Chocolate sauce",
    "price_in_cents": 500,
    "type": "TOPPING",
    "is_favorite": true
  },
  {
    "id": 8,
    "product_name": "Lemon",
    "price_in_cents": 2,
    "type": "TOPPING",
    "is_favorite": false
  }
]